File: scripts/check_architecture.py
#!/usr/bin/env python3
"""Enforce Tapid's physical source file architecture thresholds."""

import argparse
import subprocess
import sys
from pathlib import Path, PurePosixPath


REVIEW_THRESHOLD = 800
CLI_MAIN_PATH = "crates/tapid-cli/src/main.rs"
CLI_MAIN_THRESHOLD = 100
EXCEPTIONS_PATH = "docs/architecture-exceptions.txt"
MIN_RATIONALE_CHARS = 20
EXCLUDED_TEST_PARTS = frozenset({"test", "tests"})
EXCLUDED_TOP_LEVEL_TREES = frozenset({"target", "generated", "build"})
EXCLUDED_TEST_FILES = frozenset({"test.rs", "tests.rs"})


def tracked_files(root):
    result = subprocess.run(
        ["git", "ls-files", "-z"],
        cwd=root,
        check=True,
        capture_output=True,
    )
    return sorted(
        path.decode("utf-8")
        for path in result.stdout.split(b"\0")
        if path
    )


def is_production_rust(path):
    pure_path = PurePosixPath(path)
    return (
        pure_path.suffix == ".rs"
        and pure_path.name not in EXCLUDED_TEST_FILES
        and pure_path.parts[0] not in EXCLUDED_TOP_LEVEL_TREES
        and not any(part in EXCLUDED_TEST_PARTS for part in pure_path.parts)
    )


def read_exceptions(root, tracked):
    if EXCEPTIONS_PATH not in tracked:
        return {}, []

    exceptions = {}
    errors = []
    text = (root / EXCEPTIONS_PATH).read_text(encoding="utf-8")
    for line_number, raw_line in enumerate(text.splitlines(), start=1):
        line = raw_line.strip()
        if not line or line.startswith("#"):
            continue
        path, separator, rationale = line.partition("|")
        path = path.strip()
        rationale = rationale.strip()
        if not separator or not path or not rationale:
            errors.append(
                f"{EXCEPTIONS_PATH}: line {line_number} must contain a path and rationale"
            )
            continue
        if len(rationale) < MIN_RATIONALE_CHARS:
            errors.append(
                f"{EXCEPTIONS_PATH}: line {line_number} rationale must be at least "
                f"{MIN_RATIONALE_CHARS} characters"
            )
            continue
        if path in exceptions:
            errors.append(f"{EXCEPTIONS_PATH}: duplicate exception for {path}")
            continue
        exceptions[path] = rationale
    return exceptions, errors


def physical_line_count(path):
    with path.open("rb") as source_file:
        return sum(1 for _ in source_file)


def check(root):
    tracked = tracked_files(root)
    production_files = [path for path in tracked if is_production_rust(path)]
    line_counts = {
        path: physical_line_count(root / path) for path in production_files
    }
    exceptions, errors = read_exceptions(root, set(tracked))

    for path in sorted(exceptions):
        if path not in production_files:
            errors.append(
                f"{EXCEPTIONS_PATH}: exception path is not a tracked production Rust file: {path}"
            )
            continue
        threshold = CLI_MAIN_THRESHOLD if path == CLI_MAIN_PATH else REVIEW_THRESHOLD
        if line_counts[path] <= threshold:
            errors.append(
                f"{EXCEPTIONS_PATH}: exception is stale at {line_counts[path]} physical lines: {path}"
            )

    for path in production_files:
        line_count = line_counts[path]
        threshold = CLI_MAIN_THRESHOLD if path == CLI_MAIN_PATH else REVIEW_THRESHOLD
        if line_count <= threshold or path in exceptions:
            continue
        if path == CLI_MAIN_PATH:
            errors.append(
                f"{path}: {line_count} physical lines exceeds entrypoint threshold "
                f"{CLI_MAIN_THRESHOLD}; keep main.rs to argument dispatch and exit conversion "
                f"or document an exception in {EXCEPTIONS_PATH}"
            )
        else:
            errors.append(
                f"{path}: {line_count} physical lines exceeds {REVIEW_THRESHOLD}; "
                f"split the module or document an exception in {EXCEPTIONS_PATH}"
            )

    return production_files, sorted(errors)


def parse_args(argv):
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument(
        "--root",
        type=Path,
        default=Path.cwd(),
        help="repository root, default: current directory",
    )
    return parser.parse_args(argv)


def main(argv=None):
    args = parse_args(argv)
    root = args.root.resolve()
    try:
        production_files, errors = check(root)
    except (OSError, subprocess.CalledProcessError, UnicodeDecodeError) as error:
        print(f"Architecture check could not run: {error}")
        return 2

    if errors:
        print("Architecture check failed:")
        for error in errors:
            print(f"- {error}")
        return 1

    noun = "file" if len(production_files) == 1 else "files"
    print(
        f"Architecture check passed: {len(production_files)} production Rust {noun} scanned"
    )
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: tools/check_architecture.ts
#!/usr/bin/env -S node --experimental-strip-types

import { execFile } from "node:child_process";
import { lstat, readFile } from "node:fs/promises";
import { join, resolve } from "node:path";
import { promisify } from "node:util";
import { pathToFileURL } from "node:url";

const execFileAsync = promisify(execFile);
const REVIEW_RECOMMENDATION = 800;
const CLI_MAIN_PATH = "crates/tapid-cli/src/main.rs";
const CLI_MAIN_THRESHOLD = 100;
const EXCEPTIONS_PATH = "docs/architecture-exceptions.txt";
const MIN_RATIONALE_CHARS = 20;
const EXCLUDED_TEST_PARTS = new Set(["test", "tests"]);
const EXCLUDED_TOP_LEVEL_TREES = new Set(["target", "generated", "build"]);
const EXCLUDED_TEST_FILES = new Set(["test.rs", "tests.rs"]);

async function trackedFiles(root: string): Promise<string[]> {
  const { stdout } = await execFileAsync("git", ["ls-files", "-z"], { cwd: root, encoding: "utf8" });
  const existing: string[] = [];
  for (const path of stdout.split("\0").filter(Boolean)) {
    try {
      await lstat(join(root, path));
      existing.push(path);
    } catch (error) {
      if (!(error instanceof Error) || !("code" in error) || error.code !== "ENOENT") throw error;
    }
  }
  return existing.sort();
}

function isProductionRust(path: string): boolean {
  const parts = path.split("/");
  const name = parts.at(-1) ?? "";
  return name.endsWith(".rs") &&
    !EXCLUDED_TEST_FILES.has(name) &&
    !EXCLUDED_TOP_LEVEL_TREES.has(parts[0]) &&
    !parts.some((part) => EXCLUDED_TEST_PARTS.has(part));
}

async function physicalLineCount(path: string): Promise<number> {
  const bytes = await readFile(path);
  if (bytes.length === 0) return 0;
  let lines = bytes[bytes.length - 1] === 10 ? 0 : 1;
  for (const byte of bytes) if (byte === 10) lines++;
  return lines;
}

async function readExceptions(root: string, tracked: Set<string>) {
  const exceptions = new Map<string, string>();
  const errors: string[] = [];
  if (!tracked.has(EXCEPTIONS_PATH)) return { exceptions, errors };

  const text = await readFile(join(root, EXCEPTIONS_PATH), "utf8");
  for (const [index, rawLine] of text.split(/\r?\n/).entries()) {
    const line = rawLine.trim();
    if (!line || line.startsWith("#")) continue;
    const separator = line.indexOf("|");
    const path = separator < 0 ? "" : line.slice(0, separator).trim();
    const rationale = separator < 0 ? "" : line.slice(separator + 1).trim();
    const lineNumber = index + 1;
    if (!path || !rationale) {
      errors.push(`${EXCEPTIONS_PATH}: line ${lineNumber} must contain a path and rationale`);
    } else if (rationale.length < MIN_RATIONALE_CHARS) {
      errors.push(`${EXCEPTIONS_PATH}: line ${lineNumber} rationale must be at least ${MIN_RATIONALE_CHARS} characters`);
    } else if (exceptions.has(path)) {
      errors.push(`${EXCEPTIONS_PATH}: duplicate exception for ${path}`);
    } else {
      exceptions.set(path, rationale);
    }
  }
  return { exceptions, errors };
}

export async function checkArchitecture(root: string) {
  const tracked = await trackedFiles(root);
  const productionFiles = tracked.filter(isProductionRust);
  const lineCounts = new Map<string, number>();
  for (const path of productionFiles) lineCounts.set(path, await physicalLineCount(join(root, path)));

  const { exceptions, errors } = await readExceptions(root, new Set(tracked));
  const advisories: string[] = [];
  for (const path of [...exceptions.keys()].sort()) {
    if (!productionFiles.includes(path)) {
      errors.push(`${EXCEPTIONS_PATH}: exception path is not a tracked production Rust file: ${path}`);
    } else if (path !== CLI_MAIN_PATH) {
      errors.push(`${EXCEPTIONS_PATH}: exception path has no hard architecture threshold: ${path}`);
    } else if ((lineCounts.get(path) ?? 0) <= CLI_MAIN_THRESHOLD) {
      errors.push(`${EXCEPTIONS_PATH}: exception is stale at ${lineCounts.get(path)} physical lines: ${path}`);
    }
  }

  for (const path of productionFiles) {
    const lineCount = lineCounts.get(path) ?? 0;
    if (path === CLI_MAIN_PATH) {
      if (lineCount > CLI_MAIN_THRESHOLD && !exceptions.has(path)) {
        errors.push(`${path}: ${lineCount} physical lines exceeds entrypoint threshold ${CLI_MAIN_THRESHOLD}; keep main.rs to argument dispatch and exit conversion or document an exception in ${EXCEPTIONS_PATH}`);
      }
    } else if (lineCount > REVIEW_RECOMMENDATION) {
      advisories.push(`${path}: ${lineCount} physical lines exceeds the ${REVIEW_RECOMMENDATION}-line review recommendation; review cohesion, module depth, and navigability before deciding whether to split it`);
    }
  }
  return { productionFiles, advisories: advisories.sort(), errors: errors.sort() };
}

const isMain = process.argv[1] !== undefined && pathToFileURL(resolve(process.argv[1])).href === import.meta.url;
if (isMain) {
  const args = process.argv.slice(2);
  const validArgs = args.length === 0 ||
    (args.length === 2 && args[0] === "--root" && args[1].length > 0);
  if (!validArgs) {
    console.error("usage: check_architecture.ts [--root PATH]");
    process.exitCode = 2;
  } else {
    const root = resolve(args.length === 2 ? args[1] : process.cwd());
    try {
      const result = await checkArchitecture(root);
      if (result.errors.length) {
        console.log("Architecture check failed:");
        for (const error of result.errors) console.log(`- ${error}`);
        process.exitCode = 1;
      } else {
        if (result.advisories.length) {
          console.log("Architecture review recommended:");
          for (const advisory of result.advisories) console.log(`- ${advisory}`);
        }
        const noun = result.productionFiles.length === 1 ? "file" : "files";
        console.log(`Architecture check passed: ${result.productionFiles.length} production Rust ${noun} scanned`);
      }
    } catch (error) {
      console.error(`Architecture check could not run: ${error instanceof Error ? error.message : error}`);
      process.exitCode = 2;
    }
  }
}